### ISSUE TITLE
bug in nullspace removal?

### ISSUE THREAD

**Issue by amneetb** [IBAMR/IBAMR#1673]
I think there is a bug in removing the nullspace component in the solution vector here. Specifically, we are calculating L2_norm of the null vector, whereas it should be L2_norm^2

https://github.com/IBAMR/IBAMR/blob/master/src/navier_stokes/INSStaggeredHierarchyIntegrator.cpp#L1592

Vectors in `d_nul_vecs` are not normalized. Thus, in the final result there should be two L2_norm factors in the denominator 

Remove component of vector $\vec{b}$ from $\vec{a}$

$\vec{a} =  \vec{a} - \left(\frac{\vec{a} \cdot \vec{b}}{|| b||}\right) \frac{\vec{b}}{|| b ||} $

If we take a dot product of the above equation with $\vec{b}$ or $\vec{b}/|| b||$, we will get zero. 

The fix is just to remove the `std::sqrt` on line 1592.

**Comment by boyceg** [comment]
Why should it be the norm squared?

**Comment by amneetb** [comment]
My comment explains it :-)

**Comment by amneetb** [comment]
Basically it is a Gram-Schmidt process. We work with orthonormal vectors `q`.  The null vectors are not orthonormal. We need two normalization factors: first when taking a dot product, and second when using the null vector in the AXPY process.

**Comment by boyceg** [comment]
Ah, I missed that this was not just normalizing the vector.

Could / should the nullspace basis vectors be stored as normalized vectors in the first place?

**Comment by amneetb** [comment]
Yes, that would be the right thing to do. We also pass the vectors to other
libraries like petsc. Maybe petsc does normalization before using them. If
not then it is also not using the correct vectors (our fault not petsc’s).

--Amneet

**Comment by boyceg** [comment]
Agree that the current code is wrong. It should normalize by `nul_dot_nul`. But I think the better fix actually is to store the nullspace vectors as orthonormal vectors using modified Gram-Schmidt.

BTW, `MatNullSpaceCreate` does indeed request orthonormal nullspace vectors — which we are not currently providing.

**Comment by drwells** [comment]
Fixed by #1794.
